clippy_lints/src/casts/cast_lossless.rs

```diff
@@ -23,64 +25,70 @@ pub(super) fn check(
-    let sugg = opt.as_ref().map_or_else(
-        || {
-            app = Applicability::HasPlaceholders;
-            ".."
-        },
-        |snip| {
-            if should_strip_parens(cast_op, snip) {
-                &snip[1..snip.len() - 1]
-            } else {
-                snip.as_str()
-            }
-        },
-    );
-
-    // Display the type alias instead of the aliased type. Fixes #11285
-    //
-    // FIXME: Once `lazy_type_alias` is stabilized(?) we should use `rustc_middle` types instead,
-    // this will allow us to display the right type with `cast_from` as well.
-    let cast_to_fmt = if let TyKind::Path(QPath::Resolved(None, path)) = cast_to_hir.kind
-        // It's a bit annoying but the turbofish is optional for types. A type in an `as` cast
-        // shouldn't have these if they're primitives, which are the only things we deal with.
-        //
-        // This could be removed for performance if this check is determined to have a pretty major
-        // effect.
-        && path.segments.iter().all(|segment| segment.args.is_none())
-    {
-        snippet_with_applicability(cx, cast_to_hir.span, "..", &mut app)
-    } else {
-        cast_to.to_string().into()
-    };
-
-    let message = if cast_from.is_bool() {
-        format!("casting `{cast_from}` to `{cast_to_fmt}` is more cleanly stated with `{cast_to_fmt}::from(_)`")
-    } else {
-        format!("casting `{cast_from}` to `{cast_to_fmt}` may become silently lossy if you later change the type")
-    };
-
-    span_lint_and_sugg(
+    span_lint_and_then(
         cx,
         CAST_LOSSLESS,
         expr.span,
-        message,
-        "try",
-        format!("{cast_to_fmt}::from({sugg})"),
-        app,
+        format!("casts from `{cast_from}` to `{cast_to}` can be expressed infallibly using `From`"),
+        |diag| {
+            diag.help("an `as` cast can become silently lossy if the types change in the future");
+            let Some(from_snippet) = snippet_opt(cx, hygiene::walk_chain(cast_from_expr.span, expr.span.ctxt())) else {
+                return;
+            };
+            let Some(ty) = snippet_opt(cx, hygiene::walk_chain(cast_to_hir.span, expr.span.ctxt())) else {
+                return;
+            };
+            // The suggestion is to use a function call, so if the original expression
+            // has parens on the outside, they are no longer needed.
+            let from = if let ExprKind::Binary(_, _, _) = cast_from_expr.kind
+                && let Some(from_snippet) = from_snippet.strip_prefix('(')
+                && let Some(from_snippet) = from_snippet.strip_suffix(')')
+            {
+                from_snippet
+            } else {
+                &from_snippet
+            };
+            match cast_to_hir.kind {
+                TyKind::Infer => {
+                    let sugg = if cast_from_expr.precedence().order() == PREC_PREFIX {
+                        format!("({from}")
```

Comment:
🙃 I wonder why I did not add tests for this

It was meant to be `from_snippet` rather than `from` so that `PREC_PREFIX` would cover the unary op cases, but that still missed `x as u16 as _`

I didn't use it initially because `Sugg` had an issue with macro spans but fixing it was easier than I thought so it does now